src/actions/breakpoints/breakpointPositions.js

```diff
@@ -41,51 +39,72 @@ function convertToList(results, sourceId) {
+    // cases, the number of ranges here should be very small, and is quite
+    // likely to only be a single range.
+    for (const range of ranges) {
+      // Wrap infinite end positions to the next line to keep things simple
+      // and because we know we don't care about the end-line whitespace
+      // in this case.
+      if (range.end.column === Infinity) {
+        range.end.line += 1;
+        range.end.column = 0;
+      }
+
+      const bps = await client.getBreakpointPositions(source.actors[0], range);
+      for (const line in bps) {
+        results[line] = (results[line] || []).concat(bps[line]);
+      }
     }
+  } else {
+    results = await client.getBreakpointPositions(source.actors[0]);
+  }
+
+  const positions = convertToList(results, sourceId);
+  return mapLocations(positions, getState(), source, sourceMaps);
+}
 
-    let source = getSourceFromId(getState(), sourceId);
+export function setBreakpointPositions(sourceId: string) {
+  return async (thunkArgs: ThunkArgs) => {
+    const { dispatch, getState } = thunkArgs;
 
-    let results = {};
-    if (isOriginalId(sourceId)) {
-      const ranges = await sourceMaps.getGeneratedRangesForOriginal(
+    if (hasBreakpointPositions(getState(), sourceId)) {
+      return;
+    }
+
+    if (!requests.has(sourceId)) {
+      requests.set(
         sourceId,
-        source.url,
-        true
+        (async () => {
+          try {
+            dispatch({
+              type: "ADD_BREAKPOINT_POSITIONS",
+              sourceId,
+              positions: await getBreakpointPositions(sourceId, thunkArgs)
+            });
+          } finally {
+            requests.delete(sourceId);
+          }
+        })()
       );
-      sourceId = originalToGeneratedId(sourceId);
-      source = getSourceFromId(getState(), sourceId);
-
-      // Note: While looping here may not look ideal, in the vast majority of
-      // cases, the number of ranges here should be very small, and is quite
-      // likely to only be a single range.
-      for (const range of ranges) {
-        // Wrap infinite end positions to the next line to keep things simple
-        // and because we know we don't care about the end-line whitespace
-        // in this case.
-        if (range.end.column === Infinity) {
-          range.end.line += 1;
-          range.end.column = 0;
-        }
-
-        const bps = await client.getBreakpointPositions(
-          source.actors[0],
-          range
-        );
-        for (const line in bps) {
-          results[line] = (results[line] || []).concat(bps[line]);
-        }
-      }
-    } else {
-      results = await client.getBreakpointPositions(source.actors[0]);
     }
 
-    let positions = convertToList(results, sourceId);
-    positions = await mapLocations(positions, getState(), source, sourceMaps);
-    return dispatch({ type: "ADD_BREAKPOINT_POSITIONS", positions, location });
+    return requests.get(sourceId);
```

Comment:
nit: this is really just an `await` since the promise doesn't resolve to a value. :P